fondant/manifest.py

```diff
@@ -188,12 +188,22 @@ def remove_subset(self, name: str) -> None:
 
         del self._specification["subsets"][name]
 
-    def evolve(self, component_spec: FondantComponentSpec) -> "Manifest":
-        """Evolve the manifest based on the provided component spec. The resulting manifest is
-        the expected result if the current manifest is provided to the component defined by the
-        component spec."""
+    def evolve(
+        self, component_id: str, component_spec: FondantComponentSpec
+    ) -> "Manifest":
+        """Evolve the manifest based on the component spec. The resulting
+        manifest is the expected result if the current manifest is provided to the component defined
+        by the component spec."""
         evolved_manifest = self.copy()
 
+        # Update `component_id` of the metadata
+        evolved_manifest.add_metadata(key="component_id", value=component_id)
+
+        # Update index location as this is currently always rewritten
```

Comment:
@RobbeSneyders I was wondering whether the decoupling between the `evolve` method and the `Dataset` class is convenient, given that the `evolve` method defines new locations of files without actually uploading any data. e.g. here I need to update the location of the index, since the `Dataset` class will be used to write a new index.

Wouldn't it make more sense to just have a single place where both a location is defined + data is written to the cloud (i.e. both in the `Dataset` class)?

Reply:
> Otherwise we will need a strict coupling between the Manifest and Dataset classes to make sure that the latter will upload new data to new locations defined by the evolve method.

That's indeed what I'm proposing. There's a couple of advantages of splitting the `evolve` functionality out of the `Dataset` class:
- Separation of concerns, which makes it easier to follow the logic.
- The `evolve` method works purely based on a `Manifest` and a `ComponentSpec` which allows you to precalculate the manifests for a complete pipeline statically.

I see no reason we cannot include the location update in the `evolve` method.

Reply:
Updating the index location is now part of the `evolve` method